### ISSUE TITLE
command not found after start zsh

### ISSUE THREAD

**Issue by TimoFrenzel** [robbyrussell/oh-my-zsh#4771]
```
$ zsh
complete:13: command not found: compdef
```

**Comment by apjanke** [comment]
Might be a corrupted compdump file. Try doing `rm ~/*zcompdump*` and restarting zsh.

**Comment by ryanfitz** [comment]
I've also been running into this for a couple weeks now after upgrading oh-my-zsh. 

@apjanke I'm running the latest version on master, I removed all compdump files, but am still seeing `complete:13: command not found: compdef`

**Comment by mcornella** [comment]
This could also be a compinit security problem, does `compaudit` return anything?

**Comment by ryanfitz** [comment]
@mcornella compaudit returns nothing for me.

**Comment by mcornella** [comment]
Check your `$fpath` then, maybe you're redefining it somewhere in your zshrc file, or perhaps it's wrong by default in your zsh installation. Which platform are you running and how was zsh installed?

If that doesn't help, here are some other issues that might help get you in the right track: https://github.com/robbyrussell/oh-my-zsh/issues?q=%22complete+command+not+found+compdef%22+is%3Aclosed

**Comment by daneah** [comment]
This also began cropping up for me recently, particularly when I attempt to `ssh` anywhere. `compaudit` returns nothing, and removing `.zcompdump*` files has no effect. The previously-opened issues did not solve this for me, either.

I've been running zsh and oh-my-zsh successfully on OS X and AWS-flavored CentOS for quite some time and tested all recent changes to my dotfiles to rule out my configuration as the issue.

In regards to `$fpath`, I notice that `/scripts/zsh/Completion` as well as `/path/to/oh-my-zsh/completions` are present. I don't think this is the issue, but please correct me if I'm wrong!

**Comment by mcornella** [comment]
Ok I don't have any more ideas left, let's see your zshrc file.

**Comment by ryanfitz** [comment]
Here is my [zshrc](https://github.com/ryanfitz/dotfiles/blob/master/amazon-linux/zshrc).

I'm also running aws-flavored centos and osx. No issues on osx, only on aws - centos. 

```
$ zsh --version
zsh 5.0.2 (x86_64-redhat-linux-gnu)

$ echo $fpath
/home/ec2-user/.oh-my-zsh/plugins/tmux /home/ec2-user/.oh-my-zsh/plugins/npm /home/ec2-user/.oh-my-zsh/plugins/node /home/ec2-user/.oh-my-zsh/plugins/git /home/ec2-user/.oh-my-zsh/functions /home/ec2-user/.oh-my-zsh/completions /usr/share/zsh/site-functions /usr/share/zsh/5.0.2/functions
```

**Comment by daneah** [comment]
Sure thing! It's [here](https://github.com/daneah/dotfiles/blob/master/source/zshrc).

Everything in my `.zsh-custom` only defines custom environment variables and enables the zsh-syntax-highlighting plugin. [`aliases.zsh`](https://github.com/daneah/dotfiles/blob/master/source/aliases.zsh) and [`functions.zsh`](https://github.com/daneah/dotfiles/blob/master/source/functions.zsh) are sourced in `zshrc` if you'd like to take a look there, too.

```shell
$ zsh --version
zsh 5.0.7 (x86_64-apple-darwin13.4.0)

$ echo $fpath
/scripts/zsh/Completion /Users/dhillard/.zsh-custom/plugins/zsh-syntax-highlighting /Users/dhillard/.oh-my-zsh/plugins/vagrant /Users/dhillard/.oh-my-zsh/plugins/virtualenvwrapper /Users/dhillard/.oh-my-zsh/plugins/supervisor /Users/dhillard/.oh-my-zsh/plugins/rvm /Users/dhillard/.oh-my-zsh/plugins/osx /Users/dhillard/.oh-my-zsh/plugins/python /Users/dhillard/.oh-my-zsh/plugins/pip /Users/dhillard/.oh-my-zsh/plugins/npm /Users/dhillard/.oh-my-zsh/plugins/httpie /Users/dhillard/.oh-my-zsh/plugins/history-substring-search /Users/dhillard/.oh-my-zsh/plugins/history /Users/dhillard/.oh-my-zsh/plugins/git /Users/dhillard/.oh-my-zsh/plugins/docker /Users/dhillard/.oh-my-zsh/plugins/django /Users/dhillard/.oh-my-zsh/plugins/brew /Users/dhillard/.oh-my-zsh/plugins/bower /Users/dhillard/.oh-my-zsh/functions /Users/dhillard/.oh-my-zsh/completions /usr/local/share/zsh/site-functions /usr/share/zsh/site-functions /usr/share/zsh/5.0.8/functions

**Comment by mcornella** [comment]
The `compdef` function is defined in [`Completion/compinit`](https://github.com/zsh-users/zsh/blob/master/Completion/compinit#L238). You should be able to find that file from one of your `$fpath` directories. Try the following:
```zsh
for d ($fpath); do [[ -f "$d/compinit" ]] && echo $d/compinit; done
```

**Comment by daneah** [comment]
For me, this command returns `/usr/share/zsh/5.0.8/functions/compinit`

**Comment by swvjeff** [comment]
@mcornella This is also happening to me on newer Amazon Linux machines. All of my .zshrc files are unchanged. I also have a very old Amazon Linux box in which I'm not receiving this error. IIRC, I first noticed the error about month ago, but I'm not exactly sure off the top of my head.

My output for your command is: `/usr/share/zsh/5.0.2/functions/compinit`

**Comment by xVir** [comment]
Also encountered this problem...

**Comment by mcornella** [comment]
I have no idea what's going on. Your zshrc files don't show nothing suspicious, so I'd like you to generate a diagnostic dump by running `omz_diagnostic_dump` and posting the generated file as a gist.

**Comment by ryanfitz** [comment]
@mcornella here is my [omz_diagnostic_dump](https://gist.github.com/ryanfitz/2f2dbf1d50855a6a919f)

**Comment by mcornella** [comment]
Looking into it...

**Comment by mcornella** [comment]
Wait no, that won't help me diagnose this. What I wanted instead was that you run zsh in debug mode and post the output as a gist. Follow [these instructions from the wiki](https://github.com/robbyrussell/oh-my-zsh/wiki/Troubleshooting#everything-else). Sorry for the mistake! :sweat_smile:

**Comment by ryanfitz** [comment]
Here is my [omz-debug.log](https://gist.github.com/ryanfitz/b5de83abed3f060b6c4c)

**Comment by apjanke** [comment]
[This section](https://gist.github.com/ryanfitz/b5de83abed3f060b6c4c#file-omz-debug-log-L144-L221) looks like it might be the problem, or at least part of it. It's from `/etc/profile.d/aws-cli.sh` and `/usr/share/zsh/site-functions/aws_zsh_completer.sh`, which looks like a custom AWS client completion for `zsh` built on `zsh`'s `bashcompinit` portability layer, and redefining `_bash_complete` using its own logic. It calls `bashcompinit`, but does not call `compinit`. It is typically `compinit` which makes `compdef` and related functions available.

This could be that the AWS client stuff assumes that it will be called after something else has called `compinit`. When you're running OMZ, it's usually `oh-my-zsh.sh` which calls `compinit`. (In your case that's being done [down here on line 567 of that log.) But since this AWS stuff is being sourced from `/etc/profile.d`, it's run before your `~/.zshrc`.

Looks like that AWS client stuff may need to be revised if it's being run from `/etc/profile.d` ahead of the user's initialization files. Specifically, it (or something preceding it in the load order) would need to call `compinit`. That may cause a problem with OMZ or normal `~/.zshrc` files which also call `compinit`, because I think that resets the defined completions.

**Comment by mcornella** [comment]
Oh crap! You beat me to it. That's exactly the problem. I'm reproducing the log at the end for future reference.

##### Explanation:

The `aws_zsh_completer.sh` file is a part of [aws/aws-cli](https://github.com/aws/aws-cli/blob/9c75270e3d8e1355220c603b04539df2a83fbca3/bin/aws_zsh_completer.sh). Funnily enough, @blueyed (who has made significant contributions to Oh My Zsh, hence the fun part) contributed the commit that dropped the call to `compinit` because it broke completion: https://github.com/aws/aws-cli/commit/a2498c41b5bb7bbc189d3251081df1971597a510. 

Now, the really offending file is `/etc/profile.d/aws-cli.sh`, because of sourcing the completion file without running `compinit`—or, at least, sourcing it in a point in time where `compinit` should've already been called. `/etc/profile.d` is soon enough in the initialization process that `compinit` hasn't been run yet. See [all the startup files for reference](http://zsh.sourceforge.net/Guide/zshguide02.html#l9).
Also, the `aws_zsh_completer.sh` file says explicitly enough that the file should be sourced in the zshrc file:

```shell
# Typically that would be called somewhere in your .zshrc.
```

I don't know which package the `aws-cli.sh` file comes from, so I cannot submit a bug report; a quick google search only returns [this file part of a Chef cookbook](https://github.com/fabn/system_base_cookbook/blob/master/templates/default/aws-cli.sh.erb), which is correct because it runs `compinit` before, as well as [a post in Japanese about this problem](http://qiita.com/ikeisuke/items/3a010e7922795eb79d76) which suggests running `compinit` in a `~/.zshenv` file to fix this problem. 

##### Solutions:

The latter solution will not work in your case though, `compinit` will be running twice and, while you won't see the error again, the aws cli completion won't work. You could then source the `aws_zsh_completer` from within your zshrc file **but that's a horrible contraption that I particularly hate**.

I think you can disable the loading of `/etc/profile.d/aws-cli.sh` by toggling off its `readable` attribute. Here's [the relevant part of the code that loads profile.d files](https://gist.github.com/ryanfitz/b5de83abed3f060b6c4c#file-omz-debug-log-L50-L54):

```zsh
        for i in /etc/profile.d/*.sh; do
	    if [ -r "$i" ]; then
	        . $i
	    fi
        done
```

The `[ -r $file ]` code checks the readable attribute of `$file`. **So with a simple `chmod a-r /etc/profile.d/aws-cli.sh` you should be all set**. 

##### Actionable items:

For everybody that posted here with this problem, you should make sure that you have the file `/etc/profile.d/aws-cli.sh`, or that at least you are using `aws-cli` in some way. If the former, you can already apply the `chmod` solution above.

If that's not the case, [post your diagnostic dump](https://github.com/robbyrussell/oh-my-zsh/issues/4771#issuecomment-180574843) and we'll figure it out.

---------

[Relevant part of the debug log](https://gist.github.com/ryanfitz/b5de83abed3f060b6c4c#file-omz-debug-log-L139-L221). Notice the comment that says "_make sure to run `compinit` before_" :joy:
```zsh
+_src_etc_profile_d:11> . /etc/profile.d/aws-cli.sh
if [ -n "$ZSH_VERSION" ]; then
   source /usr/share/zsh/site-functions/aws_zsh_completer.sh
fi
+/etc/profile.d/aws-cli.sh:1> [ -n 5.0.2 ']'
+/etc/profile.d/aws-cli.sh:2> source /usr/share/zsh/site-functions/aws_zsh_completer.sh
# Source this file to activate auto completion for zsh using the bash
# compatibility helper.  Make sure to run `compinit` before, which should be
# given usually.
#
# % source /path/to/zsh_complete.sh
#
# Typically that would be called somewhere in your .zshrc.
#
# Note, the overwrite of _bash_complete() is to export COMP_LINE and COMP_POINT
# That is only required for zsh <= edab1d3dbe61da7efe5f1ac0e40444b2ec9b9570
#
# https://github.com/zsh-users/zsh/commit/edab1d3dbe61da7efe5f1ac0e40444b2ec9b9570
#
# zsh relases prior to that version do not export the required env variables!
#
# It is planned to write a proper zsh auto completion soon. Please talk
# to Frank Becker <fb@alien8.de>.

autoload -Uz bashcompinit
+/usr/share/zsh/site-functions/aws_zsh_completer.sh:19> autoload -Uz bashcompinit
bashcompinit -i
+/usr/share/zsh/site-functions/aws_zsh_completer.sh:20> bashcompinit -i
+bashcompinit:177> unfunction bashcompinit
+bashcompinit:178> autoload -Uz bashcompinit
+bashcompinit:179> return 0

_bash_complete() {
  local ret=1
  local -a suf matches
  local -x COMP_POINT COMP_CWORD
  local -a COMP_WORDS COMPREPLY BASH_VERSINFO
  local -x COMP_LINE="$words"
  local -A savejobstates savejobtexts

  (( COMP_POINT = 1 + ${#${(j. .)words[1,CURRENT]}} + $#QIPREFIX + $#IPREFIX + $#PREFIX ))
  (( COMP_CWORD = CURRENT - 1))
  COMP_WORDS=( $words )
  BASH_VERSINFO=( 2 05b 0 1 release )

  savejobstates=( ${(kv)jobstates} )
  savejobtexts=( ${(kv)jobtexts} )

  [[ ${argv[${argv[(I)nospace]:-0}-1]} = -o ]] && suf=( -S '' )

  matches=( ${(f)"$(compgen $@ -- ${words[CURRENT]})"} )

  if [[ -n $matches ]]; then
    if [[ ${argv[${argv[(I)filenames]:-0}-1]} = -o ]]; then
      compset -P '*/' && matches=( ${matches##*/} )
      compset -S '/*' && matches=( ${matches%%/*} )
      compadd -Q -f "${suf[@]}" -a matches && ret=0
    else
      compadd -Q "${suf[@]}" -a matches && ret=0
    fi
  fi

  if (( ret )); then
    if [[ ${argv[${argv[(I)default]:-0}-1]} = -o ]]; then
      _default "${suf[@]}" && ret=0
    elif [[ ${argv[${argv[(I)dirnames]:-0}-1]} = -o ]]; then
      _directories "${suf[@]}" && ret=0
    fi
  fi

  return ret
}

complete -C aws_completer aws
+/usr/share/zsh/site-functions/aws_zsh_completer.sh:63> complete -C aws_completer aws
+complete:1> emulate -L zsh
+complete:2> local args void cmd print remove
+complete:3> args=( -C aws_completer aws ) 
+complete:4> zparseopts -D -a void o: A: G: W: C: F: P: S: X: a b c d e f g j k u v 'p=print' 'r=remove'
+complete:6> [[ -n '' ]]
+complete:8> [[ -n '' ]]
+complete:13> compdef '_bash_complete -C aws_completer' aws
complete:13: command not found: compdef
```

**Comment by apjanke** [comment]
Cool.

We (OMZ contributors) maybe need to be on the watch out for this stuff more. Seems to me there are more Linux distros and client apps that are installing `zsh` customizations at the system level, and neither they nor OMZ are designed for composability. (Nothing wrong with that! That's where pluggable stuff useful to all users should go. By default. In a user-facing system.) IIRC, we've seen this with Ubuntu and OpenSUSE, and to some extent with OS X 10.11.

No finger-pointing or blame; just, we should be less assuming that OMZ is going on top of a stock/vanilla Zsh installation. (And terminfo database, and so on.) (Also, may be worth considering that some of this code is user-level customization lifted up to system level without review by expert `zsh` users.)

Also I'll see if I can add `profile.d` stuff to `omz_diagnostic_dump`.

**Comment by daneah** [comment]
I never considered that the printed error message was coming from my AWS box rather than from my local. `chmod`ing the `aws-cli.sh` file did the trick. Thanks for the sleuthing!

**Comment by manifestinteractive** [comment]
+1 on the `chmod a-r /etc/profile.d/aws-cli.sh` trick.  Every time I logged into my AWS box I got that error and I died a little inside.  Now it's gone and I want to give everyone a hug.

**Comment by blueyed** [comment]
Where is `/etc/profile.d/aws-cli.sh` coming from?  (`dpkg -S /etc/profile.d/aws-cli.sh` on Debian/Ubuntu)
I don't have it (on Arch Linux).

**Comment by mcornella** [comment]
I believe this might come pre-installed in AWS machines. No idea who to contact though... 

I'm closing this issue as it seems to be solved for everyone.

**Comment by blueyed** [comment]
> I believe this might come pre-installed in AWS machines.

Depends on the image then, I personally have not seen it.

While you're right this issue does not really belong here, it would still be good to properly forward it.

Somebody in this thread (affected by it) should be able to provide this information, no?

**Comment by daneah** [comment]
@blueyed This file was present on my AWS CentOS installation and I don't *believe* that I put it there. If I did inadvertently it would likely have been a result of installing the aws-cli tools.

**Comment by mcornella** [comment]
@daneah here's what the [AWS website](http://docs.aws.amazon.com/cli/latest/userguide/installing.html) says (emphasis mine):

> **The AWS CLI comes pre-installed on the Amazon Linux AMI**. Run sudo yum update after connecting to the instance to get the latest version of the package available via yum. If you need a more recent version of the AWS CLI than what is available in the Amazon updates repository, uninstall the package (sudo yum remove aws-cli) and then install using pip.

Given that this is a CentOS machine, you could do `yum whatprovides /etc/profile.d/aws-cli.sh` to check that the package in question is indeed `aws-cli`, and if it is, you can do `yum info aws-cli` to see the details of the package (and hopefully the email of the maintainer to notify of this bug).

Let me know if any of those help!

**Comment by daneah** [comment]
@mcornella Right, I meant it to mean "this seems pre-installed" :smile: sorry for the confusion.

**Comment by mcornella** [comment]
Exactly, that's what I understood haha. The quote above backs you up further, so your original intuition was right :wink: 

Let me know if you get a chance to test the commands above so we can escalate this bug to whoever is maintaining the package.Cheers!

**Comment by daneah** [comment]
`aws-cli` is officially maintained by AWS themselves, to my knowledge! Running the appropriate `chmod` did fix this issue for me.

**Comment by mcornella** [comment]
Ok you were right. Fortunately someone opened an issue a couple of days ago and I found it just now. The relevant issue is at https://github.com/aws/aws-cli/issues/1819 if you want to follow the discussion.

Thanks everyone!

**Comment by blueyed** [comment]
Thanks @mcornella!

**Comment by spanktar** [comment]
**TL;DR**
aws-cli will get updated someday, until then:

`chmod a-r /etc/profile.d/aws-cli.sh`

**Comment by acertain** [comment]
This is fixed as of aws-cli-1.11.83-1.46, pushed today.

**Comment by ccorcos** [comment]
I used the chmod solution and it worked, but now I cannot `chsh`:

```
~ ❯❯❯ sudo chsh -s /bin/zsh
Changing shell for root.
chsh: Shell not changed.
```

**Comment by mcornella** [comment]
Is this still happening folks?
